runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java

```diff
@@ -234,19 +237,42 @@ private void scheduleConsumers(ExecutorUpdate update) {
   }
 
   @Override
-  public void awaitCompletion() throws Exception {
-    VisibleExecutorUpdate update;
-    do {
+  public State awaitCompletion(Duration duration) throws Exception {
+    Instant completionTime;
+    if (duration.equals(Duration.ZERO)) {
+      completionTime = new Instant(Long.MAX_VALUE);
+    } else {
+      completionTime = Instant.now().plus(duration);
+    }
+    VisibleExecutorUpdate update = null;
+    while (Instant.now().isBefore(completionTime)
+        && (update == null || isTerminalStateUpdate(update))) {
       // Get an update; don't block forever if another thread has handled it
       update = visibleUpdates.poll(2L, TimeUnit.SECONDS);
       if (update == null && executorService.isShutdown()) {
         // there are no updates to process and no updates will ever be published because the
         // executor is shutdown
-        return;
+        return State.UNKNOWN;
       } else if (update != null && update.exception.isPresent()) {
         throw update.exception.get();
       }
-    } while (update == null || !update.isDone());
+    }
+    if (update == null || update.getNewState() == null) {
+      return executorService.isShutdown() ? State.UNKNOWN : State.RUNNING;
```

Comment:
Since we control the executorService, could we make sure it provides the state when it is shutting down?